java/src/main/native/src/DataSourceHelperJni.cpp

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/io/datasource.hpp>
+
+#include "cudf_jni_apis.hpp"
+#include "jni_utils.hpp"
+
+namespace {
+
+#define DATA_SOURCE_CLASS "ai/rapids/cudf/DataSource"
+
+jclass DataSource_jclass;
+jmethodID hostRead_method;
+jmethodID hostReadBuff_method;
+jmethodID hostReadBuffDone_method;
+jmethodID deviceRead_method;
+
+} // anonymous namespace
+
+namespace cudf {
+namespace jni {
+bool cache_data_source_jni(JNIEnv *env) {
+  jclass cls = env->FindClass(DATA_SOURCE_CLASS);
+  if (cls == nullptr) {
+    return false;
+  }
+
+  hostRead_method = env->GetStaticMethodID(cls, "hostRead", "(L" DATA_SOURCE_CLASS ";JJJ)J");
```

Comment:
Nit: Seems like it would be more straightforward on the Java end if these were member functions rather than static functions, and the JNI invoked it as a member function directly since the native code already has the object pointer.  Curious if that was considered.